File: P1.2/P1_S2/src/p1_s2/EventoDeportivo.java
package p1_s2;

import java.util.Random;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * 
 * @author Raquel Molina Reche (GH: rmr00),
 * Pablo Jiménez Jiménez (GH: pablojj1808)
 */
public class EventoDeportivo extends Evento {

    public EventoDeportivo(int ne) {
        super(ne, "DEPORTIVO");
        
        System.out.println(C.GREEN_BOLD+"Se han vendido un total de " + ne + " entradas para el evento DEPORTIVO"+C.RESET);

    }
    
    @Override
    String identificate() {
        return nombre;
    }

    @Override
    public void run() {
         try {
            System.out.println(C.GREEN_BACKGROUND+C.WHITE+"¡Comienza el evento "+nombre+"!"+C.RESET);

            Random rand = new Random();

            //Un 30% de los asistentes salen del concierto por alguna razón
            int salidas = (int)(publico.size() * 0.3);
            int num_aleatorio;

            //System.out.println("Salidas "+salidas);


            for(int i=0; i <salidas; i++ ){
                //Se realiza aleatoriamente una acción
                int accion = rand.nextInt(6);
                //System.out.println("accion "+accion);

                switch(accion){ //Se elimina aleatoriamente a esos participantes
                    case 0: //Mareos
                        num_aleatorio = rand.nextInt(publico.size());
                        System.out.println(C.GREEN+"El usuario "+ publico.get(num_aleatorio).getIdentificador() +" sale del evento deportivo porque se marea");
                        publico.remove(num_aleatorio);
                        this.sleep(1000);
                        break;

                    case 1: //No les gusta
                        num_aleatorio = rand.nextInt(publico.size());
                        System.out.println(C.GREEN+"El usuario "+ publico.get(num_aleatorio).getIdentificador() +" sale del evento deportivo porque no le está gustando");
                        publico.remove(num_aleatorio);
                        this.sleep(2000);
                        break;
                    case 2:
                        num_aleatorio = rand.nextInt(publico.size());
                        System.out.println(C.GREEN+"El usuario "+ publico.get(num_aleatorio).getIdentificador() +" sale del evento deportivo porque tiene sueño");
                        publico.remove(num_aleatorio);
                        this.sleep(3000);
                        break;
                    case 3: //Mareos
                        num_aleatorio = rand.nextInt(publico.size());
                        System.out.println(C.GREEN+"El usuario "+ publico.get(num_aleatorio).getIdentificador() +" sale del evento deportivo porque se marea");
                        publico.remove(num_aleatorio);
                        this.sleep(1000);
                        break;

                    case 4: //No les gusta
                        num_aleatorio = rand.nextInt(publico.size());
                        System.out.println(C.GREEN+"El usuario "+ publico.get(num_aleatorio).getIdentificador() +" sale del evento deportivo porque no le está gustando");
                        publico.remove(num_aleatorio);
                        this.sleep(2000);
                        break;
                    case 5:
                        num_aleatorio = rand.nextInt(publico.size());
                        System.out.println(C.GREEN+"El usuario "+ publico.get(num_aleatorio).getIdentificador() +" sale del evento deportivo porque tiene sueño");
                        publico.remove(num_aleatorio);
                        this.sleep(3000);
                        break;
                }
            }
            
            System.out.println(C.GREEN_BACKGROUND+C.WHITE+"¡Termina el evento "+nombre+"!"+C.RESET);

        } catch (InterruptedException ex) {
            Logger.getLogger(EventoConcierto.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

}

File: P1.2/P1_S2_rb/EventoDeportivo.rb
require_relative 'Evento.rb'
require_relative 'Color.rb'


class EventoDeportivo < Evento

    public_class_method :new 
    
    attr_reader :name, :n_entradas, :publico

    def initialize nombre, n_entradas, publico
        super nombre, n_entradas, publico
    end

    def empezar 
        puts "#{Colors::KGRN} Se han vendido un total de #{@n_entradas} entradas para el evento #{@name}  #{Colors::RST}"
        puts "#{Colors::FGRN}#{Colors::KWHT}Empieza el evento #{@name} #{Colors::RST}"

        #Un 30% de los asistentes salen del evento deporivo por alguna razón
        salidas = @publico.length() * 0.3

        for i in (0..salidas)
            
            #Se realiza aleatoriamente una accion
            accion = Random.new.rand(3)

            case accion
            when 0
                num_aleatorio = Random.new.rand(@publico.length)
                puts "#{Colors::KGRN}\tEl usuario #{@publico.at(num_aleatorio).id} sale del evento deporivo porque se marea #{Colors::RST}"
                @publico.delete_at(num_aleatorio)
            when 1
                num_aleatorio = Random.new.rand(@publico.length)
                puts "#{Colors::KGRN} \tEl usuario #{@publico.at(num_aleatorio).id} sale del evento deporivo porque se ha cansado #{Colors::RST}"
                @publico.delete_at(num_aleatorio)   
            when 2
                num_aleatorio = Random.new.rand(@publico.length)
                puts "#{Colors::KGRN} \tEl usuario #{@publico.at(num_aleatorio).id} sale del evento deporivo porque le están llamando por telefono #{Colors::RST}"
                @publico.delete_at(num_aleatorio) 

            end
        end

        puts "#{Colors::FGRN} #{Colors::KWHT}Termina el evento #{@name} #{Colors::RST}"

    end

    def clone
        EventoDeportivo.new @name, @n_entradas, @publico
    end    

end